devito/interfaces.py

```diff
@@ -440,44 +445,126 @@ class DenseData(TensorData):
+                self.dtype = kwargs.get('dtype', np.float32)
+                if self.shape_domain is None:
+                    error("Creating a Function requires either 'shape'"
+                          "or a 'grid' argument")
+                    raise ValueError("Unknown symbol dimensions or shape")
+            else:
+                self.shape_domain = self.grid.shape_domain
+                self.dtype = kwargs.get('dtype', self.grid.dtype)
             self.indices = self._indices(**kwargs)
-            self.dtype = kwargs.get('dtype', np.float32)
+
             self.space_order = kwargs.get('space_order', 1)
             self.initializer = kwargs.get('initializer', None)
             if self.initializer is not None:
                 assert(callable(self.initializer))
             self._first_touch = kwargs.get('first_touch', configuration['first_touch'])
             self._data_object = None
 
+            # Dynamically create notational shortcuts for space derivatives
+            for dim in self.space_dimensions:
+                # First derivative, centred
+                dx = partial(first_derivative, order=self.space_order,
+                             dim=dim, side=centered)
+                setattr(self.__class__, 'd%s' % dim.name,
+                        property(dx, 'Symbol for the first derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                # First derivative, left
+                dxl = partial(first_derivative, order=self.space_order,
+                              dim=dim, side=left)
+                setattr(self.__class__, 'd%sl' % dim.name,
+                        property(dxl, 'Symbol for the first derivative '
+                                 'wrt %s with a left stencil' % dim.name))
+
+                # First derivative, right
+                dxr = partial(first_derivative, order=self.space_order,
+                              dim=dim, side=right)
+                setattr(self.__class__, 'd%sr' % dim.name,
+                        property(dxr, 'Symbol for the first derivative '
+                                 'wrt %s with a right stencil' % dim.name))
+
+                # Second derivative
+                dx2 = partial(generic_derivative, order=2, dim=dim,
+                              width=self.space_order / 2)
+                setattr(self.__class__, 'd%s2' % dim.name,
+                        property(dx2, 'Symbol for the second derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                # Fourth derivative
+                dx4 = partial(generic_derivative, order=4, dim=dim,
+                              width=max(int(self.space_order / 2), 2))
+                setattr(self.__class__, 'd%s4' % dim.name,
+                        property(dx4, 'Symbol for the fourth derivative '
+                                 'wrt the %s dimension' % dim.name))
+
+                for dim2 in self.space_dimensions:
```

Comment:
This will create duplicates properties (e.g `dxy`/`dyx`), not really a problem just unnecessary

Reply:
Hmmm, good point. It probably will, but they should be symbolically equivalent, so I don't think we need to worry about it too much.